src/common/cockpitjson.c

```diff
@@ -967,3 +967,165 @@ cockpit_json_to_hash_table (JsonObject *object,
+{
+  gint n = json_array_get_length (array);
+  JsonArray *copy = NULL;
+
+  for (gint i = 0; i < n; i++)
+    {
+      JsonNode *element = json_array_get_element (array, i);
+      JsonNode *node = cockpit_json_walk_node (element, closure);
+
+      if (node && !copy)
+        {
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_array_sized_new (n);
+          for (gint j = 0; j < i; j++)
+            json_array_add_element (copy, json_array_dup_element (array, j));
+        }
+
+      if (copy)
+        json_array_add_element (copy, node ?: json_node_ref (element));
+    }
+
+  return copy;
+}
+
+static JsonObject *
+cockpit_json_walk_object (JsonObject  *object,
+                          WalkClosure *closure)
+{
+  JsonObject *copy = NULL;
+  JsonObjectIter iter;
+  const gchar *key;
+  JsonNode *value;
+
+  json_object_iter_init (&iter, object);
+  while (json_object_iter_next (&iter, &key, &value))
+    {
+      JsonNode *node = cockpit_json_walk_node (value, closure);
+
+      if (node && !copy)
+        {
+          const gchar *this_key = key;
+
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_object_new ();
+          json_object_iter_init (&iter, object);
+          while (json_object_iter_next (&iter, &key, &value) && key != this_key)
+            json_object_set_member (copy, key, json_node_ref (value));
+
+          /* ...now key and value will be where they were before we entered */
+          g_assert (key == this_key);
+        }
+
+      if (copy)
+        json_object_set_member (copy, key, node ?: json_node_ref (value));
+    }
+
+  return copy;
+}
+
+static JsonNode *
+cockpit_json_walk_node (JsonNode    *node,
+                        WalkClosure *closure)
+{
+  JsonNode *new_node = NULL;
+
+  switch (json_node_get_node_type (node))
+    {
+    case JSON_NODE_ARRAY:
+      {
+        JsonArray *array = json_node_get_array (node);
+        JsonArray *copy = cockpit_json_walk_array (array, closure);
+
+        if (copy)
+          {
+            new_node = json_node_new (JSON_NODE_ARRAY);
+            json_node_take_array (new_node, copy ?: json_array_ref (array));
+          }
+
+        break;
+      }
+
+    case JSON_NODE_OBJECT:
+      {
+        JsonObject *object = json_node_get_object (node);
+        JsonObject *copy = cockpit_json_walk_object (object, closure);
+
+        if (copy)
+          {
+            new_node = json_node_new (JSON_NODE_OBJECT);
+            json_node_take_object (new_node, copy ?: json_object_ref (object));
+          }
+
+        break;
+      }
+
+    default:
+      {
+        new_node = (* closure->callback) (node, closure->user_data);
+        break;
+      }
+    }
+
+  return new_node;
+}
+
+/**
+ * cockpit_json_walk:
```

Comment:
This API seems rather nice, and also very generic -- do you plan to submit that to json-glib?

Reply:
I wasn't planning on it, but I could file a PR and see what they say, I guess.  I imagine they'd want an awful lot more testing than we do here...